### PR TITLE
dns resolver: add options to initialize c-ares with custom timeout and attempts

### PR DESCRIPTION
**Pull request by agrawroh**
## Background

We recently saw a small number of DNS queries getting timed out. There is currently no option for us to [initialize c-ares](https://c-ares.org/ares_init_options.html) by passing in a custom [timeout](https://c-ares.org/ares_init_options.html#:~:text=c%2Dares%201.5.2.-,ARES_OPT_TIMEOUTMS,-int%20timeout%3B) and [tries](https://c-ares.org/ares_init_options.html#:~:text=c%2Dares%201.5.2.-,ARES_OPT_TRIES,-int%20tries%3B) config options.

## Changes

In this PR, we are adding two new options under `DnsResolverOptions` for configuring custom timeouts and tries while resolving DNS queries. Custom timeouts could be configured by specifying `dns_resolver_query_timeout` option and custom tries could be configured by specifying `dns_resolver_query_tries` config option.

Commit Message: dns resolver: add options to initialize c-ares with custom timeout and attempts
Additional Description: add two new options called `dns_resolver_query_timeout` and `dns_resolver_query_tries` for initializing c-ares with custom timeout and tries.
Risk Level: Low
Testing: Added Unit/Integration Tests
Docs Changes: Added
Release Notes: Added
Platform Specific Features: N/A
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/config/core/v3/resolver.proto [hunk 1]
```diff
@@ -22,6 +22,14 @@ message DnsResolverOptions {
 
   // Do not use the default search domains; only query hostnames as-is or as aliases.
   bool no_default_search_domain = 2;
+
+  // The initial response timeout in milliseconds given to a name server to reply to a query on the first attempt.
+  // After the initial attempt, the timeout algorithm becomes more intricate, but remains proportionate to this value.
+  // The default is 2000 ms (2 seconds).
+  uint32 dns_resolver_query_timeout_ms = 3 [(validate.rules).uint32 = {gt: 0}];
```
Comment (lizan) [inline api/envoy/config/core/v3/resolver.proto:29]: use `google.protobuf.Duration` for timeouts.

File: api/envoy/config/core/v3/resolver.proto [hunk 2]
```diff
@@ -22,6 +22,14 @@ message DnsResolverOptions {
...
+
+  // The number of attempts the resolver will make to contact the name server before giving up. The default is 3 tries.
+  uint32 dns_resolver_query_tries = 4 [(validate.rules).uint32 = {gt: 0}];
```
Comment (lizan) [inline api/envoy/config/core/v3/resolver.proto:32]: When default is not proto3 default, use wrapped scalar types (i.e. UInt32Value).

**Review by lizan — COMMENTED** [review]
API review

File: api/envoy/config/core/v3/resolver.proto (hunk 1 above)
Comment (agrawroh) [reply api/envoy/config/core/v3/resolver.proto]: Changed, Thanks!

File: api/envoy/config/core/v3/resolver.proto (hunk 2 above)
Comment (agrawroh) [reply api/envoy/config/core/v3/resolver.proto]: Changed, Thanks!

**Comment by agrawroh** [comment]
@lizan Whenever you get a chance, could you please take another look at the API?

**Comment by agrawroh** [comment]
@alyssawilk @yanavlasov Could you please take a look whenever you have some free time?

**Review by alyssawilk — COMMENTED** [review]
Thanks for this change - I think these will be useful knobs to have! Just have some thoughts about how we expose them to make sure we're consistently applying API
/wait

File: source/extensions/network/dns_resolver/cares/dns_impl.cc [hunk 3]
```diff
@@ -78,6 +78,20 @@ absl::optional<std::string> DnsResolverImpl::maybeBuildResolversCsv(
 DnsResolverImpl::AresOptions DnsResolverImpl::defaultAresOptions() {
   AresOptions options{};
 
+  // If the timeout is provided in the config, set the timeout to the provided value.
+  if (dns_resolver_options_.has_dns_resolver_query_timeout()) {
```
[the remark this answers is not in the document]
Comment (agrawroh) [reply source/extensions/network/dns_resolver/cares/dns_impl.cc:82]: Ah, that make sense. Let me see if these could be applied more generally to other resolvers or I'd move them to c-ares specific config. Thank you for taking a look!

**Review by lizan — COMMENTED** [review]
/lgtm api

**Pull request closed without merging**